File: src/main/java/org/kite9/diagram/visualization/compaction/position/GridCellPositionCompactionStep.java
package org.kite9.diagram.visualization.compaction.position;

import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import org.kite9.diagram.model.Connected;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.position.Dimension2D;
import org.kite9.diagram.model.position.Layout;
import org.kite9.diagram.model.position.RectangleRenderingInformation;
import org.kite9.diagram.visualization.compaction.Compaction;
import org.kite9.diagram.visualization.compaction.CompactionStep;
import org.kite9.diagram.visualization.compaction.Compactor;
import org.kite9.diagram.visualization.compaction.Embedding;
import org.kite9.diagram.visualization.orthogonalization.Dart;
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization;

/**
 * Sets the details of the cell positions in the parent container (Layout=GRID).
 */
public class GridCellPositionCompactionStep implements CompactionStep {

	public void compact(Compaction c, Embedding r, Compactor cr) {
		if (r.isTopEmbedding()) {
			Set<Container> gridContainers = createTopElementSet(c.getOrthogonalization());
			
			for (Container de : gridContainers) {
				RectangleRenderingInformation rri = de.getRenderingInformation();
				setPostions(rri, de.getContents());
			}
		}
	}
	
	private void setPostions(RectangleRenderingInformation pri, List<DiagramElement> contents) {
		if (contents == null) {
			pri.setCellXPositions(new double[] {});
			pri.setCellYPositions(new double[] {});
		} else {
			Dimension2D parent = pri.getPosition();
			Set<Double> xs = new HashSet<>();
			Set<Double> ys = new HashSet<>();
			contents.stream()
				.filter(c -> c instanceof Connected)
				.map(c -> c.getRenderingInformation()).forEach(rri -> {
				Dimension2D p = rri.getPosition();
				Dimension2D s = rri.getSize();
				xs.add(p.getWidth() - parent.getWidth());
				xs.add(p.getWidth() + s.getWidth() - parent.getWidth());
				ys.add(p.getHeight() - parent.getHeight());
				ys.add(p.getHeight() + s.getHeight() - parent.getHeight());
			});
			double[] xPositions = xs.stream().mapToDouble(d -> (double) d).sorted().toArray();
			double[] yPositions = ys.stream().mapToDouble(d -> (double) d).sorted().toArray();
			pri.setCellXPositions(xPositions);
			pri.setCellYPositions(yPositions);
		}
	}

	private Set<Container> createTopElementSet(Orthogonalization c) {
		Set<Container> out  = new LinkedHashSet<Container>();
		for (Dart e : c.getAllDarts()) {
			for (DiagramElement de : e.getDiagramElements().keySet()) {
				if (de instanceof Container) {
					if (((Container)de).getLayout() == Layout.GRID) {
						out.add((Container) de);
					}
				}
			}
		}	
		return out;
	}

}

File: src/main/kotlin/org/kite9/diagram/visualization/compaction/position/GridCellPositionCompactionStep.kt
package org.kite9.diagram.visualization.compaction.position

import org.kite9.diagram.model.Connected
import org.kite9.diagram.model.Container
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.position.Layout
import org.kite9.diagram.model.position.RectangleRenderingInformation
import org.kite9.diagram.model.position.RenderingInformation
import org.kite9.diagram.visualization.compaction.Compaction
import org.kite9.diagram.visualization.compaction.CompactionStep
import org.kite9.diagram.visualization.compaction.Compactor
import org.kite9.diagram.visualization.compaction.Embedding
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization

/**
 * Sets the details of the cell positions in the parent container (Layout=GRID).
 */
class GridCellPositionCompactionStep : CompactionStep {

    override fun compact(c: Compaction, r: Embedding, cr: Compactor) {
        if (r.isTopEmbedding) {
            val gridContainers = createTopElementSet(c.getOrthogonalization())
            for (de in gridContainers) {
                val rri = de.getRenderingInformation()
                setPostions(rri, de.getContents())
            }
        }
    }

    private fun setPostions(pri: RectangleRenderingInformation, contents: List<DiagramElement>) {
        if (contents == null) {
            pri.cellXPositions = doubleArrayOf()
            pri.cellYPositions = doubleArrayOf()
        } else {
            val parent = pri.position
            val xs: MutableSet<Double> = HashSet()
            val ys: MutableSet<Double> = HashSet()
            contents.stream()
                .filter { c: DiagramElement? -> c is Connected }
                .map { c: DiagramElement -> c.getRenderingInformation() }.forEach { rri: RenderingInformation ->
                    val p = rri.position
                    val s = rri.size
                    xs.add(p!!.w - parent!!.w)
                    xs.add(p.w + s!!.w - parent.w)
                    ys.add(p.h - parent.h)
                    ys.add(p.h + s.h - parent.h)
                }
            val xPositions = xs.stream().mapToDouble { d: Double -> d }.sorted().toArray()
            val yPositions = ys.stream().mapToDouble { d: Double -> d }.sorted().toArray()
            pri.cellXPositions = xPositions
            pri.cellYPositions = yPositions
        }
    }

    private fun createTopElementSet(c: Orthogonalization): Set<Container> {
        val out: MutableSet<Container> = LinkedHashSet()
        for (e in c.getAllDarts()) {
            for (de in e.getDiagramElements().keys) {
                if (de is Container) {
                    if (de.getLayout() === Layout.GRID) {
                        out.add(de)
                    }
                }
            }
        }
        return out
    }
}